File: wot-dijkstra/src/main/java/org/pgpainless/wot/dijkstra/sq/CertificationSet.java
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.annotation.Nonnull;

/**
 * A {@link CertificationSet} is a set of {@link Certification Certifications} made by the same issuer, on the same
 * target certificate.
 * In some sense, a {@link CertificationSet} can be considered an edge in the web of trust.
 */
public final class CertificationSet {

    private final CertSynopsis issuer;
    private final CertSynopsis target;

    private final Map<Optional<String>, List<Certification>> certifications;

    /**
     * Create an empty {@link CertificationSet}.
     *
     * @param issuer issuer
     * @param target target
     * @return empty set
     */
    public static CertificationSet empty(CertSynopsis issuer, CertSynopsis target) {
        return new CertificationSet(issuer, target, new HashMap<>());
    }

    public static CertificationSet fromCertification(Certification certification) {
        CertificationSet set = CertificationSet.empty(certification.getIssuer(), certification.getTarget());
        set.add(certification);
        return set;
    }

    private CertificationSet(CertSynopsis issuer,
                             CertSynopsis target,
                             Map<Optional<String>, List<Certification>> certifications) {
        this.issuer = issuer;
        this.target = target;
        this.certifications = new HashMap<>(certifications);
    }

    public CertSynopsis getIssuer() {
        return issuer;
    }

    public CertSynopsis getTarget() {
        return target;
    }

    public Map<Optional<String>, List<Certification>> getCertifications() {
        // Copy to avoid side effects
        Map<Optional<String>, List<Certification>> copy = new HashMap<>();
        for (Optional<String> key : certifications.keySet()) {
            copy.put(key, new ArrayList<>(certifications.get(key)));
        }
        return copy;
    }

    /**
     * Merge this {@link CertificationSet} with another instance.
     * After the operation, this will contain {@link Certification Certifications} from both sets.
     *
     * @param other other {@link CertificationSet}
     */
    public void merge(@Nonnull CertificationSet other) {
        if (other == this) {
            return;
        }

        if (!issuer.getFingerprint().equals(other.issuer.getFingerprint())) {
            throw new IllegalArgumentException("Issuer fingerprint mismatch.");
        }

        if (!target.getFingerprint().equals(other.target.getFingerprint())) {
            throw new IllegalArgumentException("Target fingerprint mismatch.");
        }

        for (Map.Entry<Optional<String>, List<Certification>> entry : other.certifications.entrySet()) {
            for (Certification certification : entry.getValue()) {
                add(certification);
            }
        }
    }

    /**
     * Add a {@link Certification} into this {@link CertificationSet}.
     *
     * @param certification certification
     */
    public void add(@Nonnull Certification certification) {
        if (!issuer.getFingerprint().equals(certification.getIssuer().getFingerprint())) {
            throw new IllegalArgumentException("Issuer fingerprint mismatch.");
        }
        if (!target.getFingerprint().equals(certification.getTarget().getFingerprint())) {
            throw new IllegalArgumentException("Target fingerprint mismatch.");
        }

        List<Certification> certificationsForUserId = certifications.get(certification.getUserId());
        // noinspection Java8MapApi
        if (certificationsForUserId == null) {
            certificationsForUserId = new ArrayList<>();
            certifications.put(certification.getUserId(), certificationsForUserId);
        }
        // TODO: Prevent duplicates, only keep newest timestamped sig?
        certificationsForUserId.add(certification);
    }

    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<Optional<String>, List<Certification>> entry : certifications.entrySet()) {
            for (Certification certification : entry.getValue()) {
                sb.append(certification).append('\n');
            }
        }
        return sb.toString();
    }
}

File: wot-dijkstra/src/main/kotlin/org/pgpainless/wot/dijkstra/sq/CertificationSet.kt
// SPDX-FileCopyrightText: 2023 Paul Schaub <vanitasvitae@fsfe.org>
//
// SPDX-License-Identifier: Apache-2.0

package org.pgpainless.wot.dijkstra.sq

/**
 * A [CertificationSet] is a set of [Certifications][Certification] made by the same issuer, on the same
 * target certificate.
 * In some sense, a [CertificationSet] can be considered an edge in the web of trust.
 *
 * @param issuer synopsis of the certificate that issued the [Certifications][Certification]
 * @param target synopsis of the certificate that is targeted by the [Certifications][Certification]
 * @param certifications [Map] keyed by user-ids, whose values are [Lists][List] of
 * [Certifications][Certification] that are calculated over the key user-id. Note, that the key can also be null for
 * [Certifications][Certification] over the targets primary key.
 */
class CertificationSet(
        val issuer: CertSynopsis,
        val target: CertSynopsis,
        certifications: Map<String?, List<Certification>>) {

    init {
        certifications.forEach { (t, u) -> _certifications[t] = u.toMutableList() }
    }

    private val _certifications: MutableMap<String?, MutableList<Certification>> = mutableMapOf()
    val certifications: Map<String?, List<Certification>>
        get() = _certifications.toMutableMap()

    companion object {

        /**
         * Create an empty [CertificationSet].
         *
         * @param issuer the certificate that issued the [Certifications][Certification].
         * @param target the certificate that is targeted by the [Certifications][Certification].
         */
        @JvmStatic
        fun empty(issuer: CertSynopsis, target: CertSynopsis): CertificationSet {
            return CertificationSet(issuer, target, HashMap())
        }

        /**
         * Create a [CertificationSet] from a single [Certification].
         *
         * @param certification certification
         */
        @JvmStatic
        fun fromCertification(certification: Certification): CertificationSet {
            val set = empty(certification.issuer, certification.target)
            set.add(certification)
            return set
        }
    }

    /**
     * Merge the given [CertificationSet] into this.
     * This method copies all [Certifications][Certification] from the other [CertificationSet] into [certifications].
     *
     * @param other [CertificationSet] with the same issuer fingerprint and target fingerprint as this object.
     */
    fun merge(other: CertificationSet) {
        if (other == this) {
            return
        }

        require(issuer.fingerprint == other.issuer.fingerprint) { "Issuer fingerprint mismatch." }
        require(target.fingerprint == other.target.fingerprint) { "Target fingerprint mismatch." }

        for (userId in other.certifications.keys) {
            for (certification in other.certifications[userId]!!) {
                add(certification)
            }
        }
    }

    /**
     * Add a single [Certification] into this objects [certifications].
     * Adding multiple [Certifications][Certification] for the same datum, but with different creation times results in
     * only the most recent [Certification(s)][Certification] to be preserved.
     *
     * @param certification [Certification] with the same issuer fingerprint and target fingerprint as this object.
     */
    fun add(certification: Certification) {
        require(issuer.fingerprint == certification.issuer.fingerprint) { "Issuer fingerprint mismatch." }
        require(target.fingerprint == certification.target.fingerprint) { "Target fingerprint mismatch." }

        var certificationsForUserId: MutableList<Certification>? = _certifications[certification.userId]
        if (certificationsForUserId == null) {
            certificationsForUserId = ArrayList()
            _certifications[certification.userId] = certificationsForUserId
        }

        if (certificationsForUserId.isEmpty()) {
            certificationsForUserId.add(certification)
            return
        }

        val existing = certificationsForUserId[0]
        if (existing.creationTime.before(certification.creationTime)) {
            certificationsForUserId.clear() // throw away older certifications
            certificationsForUserId.add(certification)
        }
        // else this certification is older, so ignore
    }

    override fun toString(): String {
        return certifications.map { it.value }.flatten().joinToString("\n")
    }
}